### PR TITLE
Vdi io path

### PR DESCRIPTION
**Pull request by Wescoeur**
@beshleman Can you check if the grammar is correct please? This branch is directly connected to https://github.com/xcp-ng/xcp-ng-org/pull/49, so can you approve or comment in the previous PR? When you have time. :wink: 

@Nambrok You can complete this PR if you have things or improvements to add in the doc. :slightly_smiling_face:

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+
+:::tip
+If you want more info concerning the persistent feature: <https://xenproject.org/2012/11/23/improving-block-protocol-scalability-with-persistent-grants/>
+
+The persistent grants are not used in `tapdisk`.
+:::
+
+3. The grant reference is then added to the ring, and the backend is notified.
+
+4. In tapdisk when the event channel is notified, the request is read and the guest segments are copied into a local buffer using `iocl` with a `IOCTL_GNTDEV_GRANT_COPY` request. So before writing to the VHD file we must have to **make a copy** of the data. Another possible solution is to use the `IOCTL_GNTDEV_MAP_GRANT_REF` + a `mmap` call to avoid a copy, but **it is not necessarily faster**.
```
Comment (beshleman) [inline docs/architecture.md:399]: ~~iocl~~ ioctl

we must ~~have to~~ make a copy

**Review by beshleman — CHANGES_REQUESTED** [review]

File: docs/architecture.md [hunk 2]
```diff
@@ -17,3 +17,387 @@ For each read/write in the VM disk, requests pass through an emulated driver, th
 
 The process described above is used for HVMs and also for PV guests (at startup, PV drivers are not loaded).
 After starting a PV guest, the emulated driver in the VM is replaced by `blkfront` (a PV driver) which allows to communicate directly with `tapdisk` using a protocol: `blkif`; `blktap` and `qemu-dm` then become useless to handle devices requests. Note that system calls are used with two drivers: `eventchn dev` and `gntdev` to map VM memory pages in the user space of the host. Thus a shared ring can be used to receive requests directly from `tapdisk` in host user space instead of using the kernel space.
+
+## Components for VDI I/O
+
+### XenStore
+
+XenStore is a centralized database residing on `Dom0`, it is accessible using a `UNIX socket` or with `XenBus` + `ioctl`. It's also a hierarchical namespace allowing to read/write, enumerate a virtual directory and to be notified when a change is observed via `watches`. Each domain has its own path in the store.
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply docs/architecture.md:25]: I will use your last suggestion. :wink: But:
> I would change "allowing to" to "allowing to"

What did you mean? :sweat_smile:

File: docs/architecture.md [hunk 3]
```diff
@@ -17,3 +17,387 @@ For each read/write in the VM disk, requests pass through an emulated driver, th
...
+
+A PV driver can interact with the `XenStore` to get or set configuration, it's not used for large data exchanges. The store is mainly used during PV driver negociation via `XenBus`.
```
[the remark this answers is not in the document]
Comment (Wescoeur) [reply docs/architecture.md:27]: :+1: When the french word "négociation" is really close to its English translation... :sweat_smile:

File: docs/architecture.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (beshleman) [reply docs/architecture.md]: Lol!  I fixed my comment, my syntax rendered into bad html haha

File: docs/architecture.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (beshleman) [reply docs/architecture.md]: I figured!

**Comment by olivierlambert** [comment]
What should we do with this PR? @Wescoeur what's left to merge it?

**Pull request merged**